f3rm_robot/initial_proposals.py

```diff
@@ -0,0 +1,128 @@
+"""
+Utilities for generating initial proposals.
+"""
+from typing import Tuple
+
+import numpy as np
+import torch
+from jaxtyping import Float
+from pytorch3d.ops import knn_points
+
+
+class NoProposalsError(Exception):
+    """Error raised when there are no proposals due to the masking."""
+
+    pass
+
+
+def dense_voxel_grid(
+    min_bounds: Tuple[float, float, float], max_bounds: Tuple[float, float, float], voxel_size: float
+) -> Float[torch.Tensor, "n_x n_y n_z 3"]:
+    """Create a dense voxel grid between min and max bounds with given voxel size."""
+    voxel_grid = torch.meshgrid(
+        *(torch.arange(min_bound, max_bound, voxel_size) for min_bound, max_bound in zip(min_bounds, max_bounds)),
+        indexing="ij",
+    )
+    voxel_grid = torch.stack(voxel_grid, dim=-1)
+    return voxel_grid
+
+
+def otsu_mask(voxel_sims: Float[torch.Tensor, "n"], num_bins: int = 100) -> Tuple[Float[torch.Tensor, "n"], float]:
+    """
+    Compute mask based on Otsu's method - i.e., maximize between-class variance.
+    Returns the mask and threshold value.
+    """
+    # Normalize the similarities
+    sim_min, sim_max = voxel_sims.min(), voxel_sims.max()
+    voxel_sims_norm = (voxel_sims - sim_min) / (sim_max - sim_min)
+
+    # Compute the histogram of the similarities
+    threshold_vals = torch.linspace(0, 1, num_bins).to(voxel_sims.device)
+    hist = torch.histc(voxel_sims_norm, bins=num_bins, min=0, max=1)
+    probs = hist / torch.sum(hist)
+
+    # Compute cumulative sum and mean
+    cum_sum = torch.cumsum(probs, dim=0)
+    cum_mean = torch.cumsum(probs * threshold_vals, dim=0)
+
+    # Calculate the between-class variance for all possible threshold values
+    variance = ((cum_mean[-1] * cum_sum - cum_mean) ** 2) / (cum_sum * (1 - cum_sum) + 1e-9)
+
+    # Find threshold that maximizes the between-class variance
+    threshold = threshold_vals[torch.argmax(variance)]
+    mask = (voxel_sims_norm >= threshold).squeeze()
+    return mask, threshold
+
+
+def marching_cubes_mask(
+    alpha: Float[torch.Tensor, "n_x n_y n_z"],
+    iso_value: float,
+    min_bounds: Tuple[float, float, float],
+    max_bounds: Tuple[float, float, float],
+) -> Float[torch.Tensor, "n 3"]:
+    """
+    Use marching cubes, so we only extract the surface of a voxel grid to reduce the number of proposals.
+    Expects alpha to be a 3D tensor, not flattened.
+
+    Returns:
+        alpha_voxel_grid: the vertices of the mesh, **flattened** as a (n, 3) tensor.
+    """
+    from mcubes import marching_cubes
+    from trimesh import Trimesh
+
+    assert alpha.dim() == 3, f"alpha should be 3D, not {alpha.shape}"
+
+    # We use mcubes as it's sufficiently fast (<0.02s). pytorch3d and kaolin are painful to deal with.
+    verts, faces = marching_cubes(alpha.cpu().numpy(), iso_value)
+
+    # Create a Trimesh, then scale and translate back to the original world coordinates
+    mesh = Trimesh(verts, faces)
```

Comment:
Trimesh sometimes has issue reading larger pointclounds. The way it parses the ply files tend to fail when trying to detect the chartset.

Reply:
Right that's good to know. We're initializing meshes directly from their numpy arrays here so it should be ok.